src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs

```diff
@@ -0,0 +1,1144 @@
+                : base(connection)
+            {
+            }
+
+            public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            {
+                if (buffer == null)
+                {
+                    throw new ArgumentNullException(nameof(buffer));
+                }
+
+                if (offset < 0 || offset > buffer.Length)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(offset));
+                }
+
+                if (count < 0 || count > buffer.Length - offset)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(count));
+                }
+
+                return _connection.WriteAsync(buffer, offset, count, cancellationToken);
+            }
+
+            public override Task FlushAsync(CancellationToken cancellationToken)
+            {
+                return _connection.FlushAsync(cancellationToken);
+            }
+
+            public override Task FinishAsync(CancellationToken cancellationToken)
+            {
+                _connection = null;
+                return Task.CompletedTask;
+            }
+        }
+
+        public HttpConnection(
+            HttpConnectionPool pool, 
+            HttpConnectionKey key, 
+            Stream stream, 
+            TransportContext transportContext, 
+            bool usingProxy)
+        {
+            _pool = pool;
+            _key = key;
+            _stream = stream;
+            _transportContext = transportContext;
+            _usingProxy = usingProxy;
+
+            _sb = new StringBuilder();
+
+            _writeBuffer = new byte[BufferSize];
+            _writeOffset = 0;
+
+            _readBuffer = new byte[BufferSize];
+            _readLength = 0;
+            _readOffset = 0;
+
+            _pool.AddConnection(this);
+        }
+
+        public HttpConnectionKey Key
+        {
+            get { return _key; }
+        }
+
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                _disposed = true;
+
+                _stream.Dispose();
+            }
+        }
+
+        private async Task<HttpResponseMessage> ParseResponseAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            HttpResponseMessage response = new HttpResponseMessage();
+            response.RequestMessage = request;
+
+            if (await ReadCharAsync(cancellationToken) != 'H' ||
+                await ReadCharAsync(cancellationToken) != 'T' ||
+                await ReadCharAsync(cancellationToken) != 'T' ||
+                await ReadCharAsync(cancellationToken) != 'P' ||
+                await ReadCharAsync(cancellationToken) != '/' ||
+                await ReadCharAsync(cancellationToken) != '1' ||
+                await ReadCharAsync(cancellationToken) != '.' ||
+                await ReadCharAsync(cancellationToken) != '1')
```

Comment:
@mconnew, while I'm sure there are ways this can be improved, ReadCharAsync is returning a ValueTask, not Task: synchronous completion is allocation-free.